reasoner/planner/RecursivePorPlanner.java

```diff
@@ -0,0 +1,361 @@
+import com.vaticle.typedb.core.traversal.TraversalEngine;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.Stack;
+
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public class RecursivePorPlanner extends ReasonerPlanner {
+    // Inaccuracies:
+    //      retrieval costs are treated the same as reasoning-overhead cost (Both approximated as the number of answers retrieved for all variables)
+    //      Excess calls are not penalized, since the scaling factor is capped at one (Solve using (calls + scaling-factor * rec-cost?) )
+    //      Acyclic dependencies are counted once for each occurence
+    //          (Can be fixed by including the optimal plans for the bound in the globalPlan chosenSummaries - The scaling factors will be capped at 1 then)
+    // So far, these are acceptable because the difference in cost is some constant factor, and our aim is to avoid horrible plans.
+    //
+    //      !!! The cost does not depend on the binding mode !!! because of the formulation - Handle this with connectedness restriction when generating orders?
+
+    final AnswerCountEstimator answerCountEstimator;
+    private final Map<CallKey, Set<OrderingSummary>> candidateOrderings;
+    private final ConjunctionSummarizer conjunctionSummarizer;
+
+    public RecursivePorPlanner(TraversalEngine traversalEng, ConceptManager conceptMgr, LogicManager logicMgr) {
+        super(traversalEng, conceptMgr, logicMgr);
+        this.conjunctionSummarizer = new ConjunctionSummarizer(logicMgr);
+        this.answerCountEstimator = new AnswerCountEstimator(logicMgr, traversalEng.graph(), this.conjunctionSummarizer);
+        this.candidateOrderings = new HashMap<>();
+    }
+
+    private Set<Variable> consideredVariables(Resolvable<?> resolvable) {
+        return iterate(resolvable.variables()).filter(Variable::isThing).toSet();
+    }
+
+    @Override
+    Plan computePlan(ResolvableConjunction conjunction, Set<Variable> inputBounds) {
+        CallKey callKey = new CallKey(conjunction, inputBounds);
+        recursivelySummarizeOrderings(callKey);
+        computeOrderingsForCycles(callKey);
+        assert planCache.getIfPresent(new Pair<>(conjunction, inputBounds)) != null;
+        return planCache.getIfPresent(new Pair<>(conjunction, inputBounds));
+    }
+
+    private void computeOrderingsForCycles(CallKey callKey) {
+        // This should be called only from conjunctions not on cycle-paths or at the lasso-knot FOR THE CURRENT LASSO ROPE
+        Map<CallKey, OrderingSummary> chosenSummaries = new HashMap<>();
+        Set<CallKey> pendingKeys = new HashSet<>();
+        pendingKeys.add(callKey);
+        Pair<List<OrderingSummary>, Double> bestPlan = cyclePlanSearch(callKey, pendingKeys, chosenSummaries);
```

Comment:
Since the plans of conjunctions in a cycle affect each other, they have to be optimised together.
cyclePlanSearch walks the rule-graph, backtracking over candidate orderings[1] at each level and generating the graph based on concludables in the chosen ordering.

[1] - Some orderings have been pruned - When multiple orderings have the same input-bounds and the same bindings for each cyclic-concludable, we only need to consider the one with the lowest cost.